DSpace/dspace-angular
# Browse by Controlled Vocabulary Page

## References
Rest PR: https://github.com/DSpace/DSpace/pull/8826
Contract PR: https://github.com/DSpace/RestContract/pull/225

## Description
This PR adds a browse page for controlled vocabularies. Which vocabulary to use is dynamically configurable from the backend.

## Instructions for Reviewers
**Changes**
- Took modal-logic out of VocabularyTreeview and put it in its own component: VocabularyTreeviewModal. (DynamicOnebox and SearchHierarchyFilter now call this component.)
- Refactored VocabularyTreeviewNodeModel, VocabularyTreeviewComponent and VocabularyTreeviewService to keep track of currently selected items and the `isSelected`-state of a checkbox.
- Added a `multiSelect`-input to VocabularyTreeviewComponent to control whether checkboxes should be displayed or not. Also added an extra output to emit deselecting-events.
- Added new BrowseDefinition models which correspond to the new `browseTypes` on the REST-side (valueList, flatBrowse and hierarchicalBrowse).
- Added new BrowseDefinitionRestRequst, BrowseResponseParsingService and BrowseDefinitionFindAllDataImpl to be able to retrieve data from the `browses` REST-endpoint.
- Added a (themed) BrowseByTaxonomyPage, which displays a VocabularyTreeview, allows the user to select/deselect multiple items and then redirects to the search-page with the selected items as search filters.
- Added an extra "Browse by Subject Category"-link to "All of DSpace". (This is automatically generated.)
- Added new tests for VocabularyTreeview, BrowseByTaxonomyPage and BrowseResponseParsingService, fixed other tests where necessary.  

**Review instructions**
- Verify the other browse pages still work as before.
- Go to the new browse page ("By Subject Category" under "All of DSpace")
- Verify the search bar and reset button work as expected.
- Select a couple of subjects, click "Browse" and verify that you are redirected to the search-page. Verify that your chosen subjects show up in the URL and are checked in the Subject sidebar facet.
- Select and deselect a couple of subjects. After clicking "Browse", verify that the subjects you deselected don't show up in the URL and aren't checked in the Subject sidebar facet.
- Verify the original use cases of VocubalaryTreeview:
    - Verify the modal works as before when choosing a subject during item submission (**config**: [input-type of subject](https://github.com/DSpace/DSpace/blob/main/dspace/config/submission-forms.xml#L197) in backend should be set to `onebox`)
    - Verify the modal works as before when choosing a subject in the Subject sidebar facet. (**config**: make sure to enable [vocabularies](https://github.com/DSpace/dspace-angular/blob/main/config/config.example.yml#L368) in config.yml)

## Checklist
